struct.c

```diff
@@ -349,6 +376,113 @@ setup_struct(VALUE nstr, VALUE members)
     return nstr;
 }
 
+static VALUE
+setup_data(VALUE nstr, VALUE members)
+{
+    long i, len;
+
+    members = struct_set_members(nstr, members);
+
+    rb_define_alloc_func(nstr, struct_alloc);
+    rb_define_singleton_method(nstr, "new", rb_data_s_new, -1);
+    rb_define_singleton_method(nstr, "[]", rb_data_s_new, -1);
+    rb_define_singleton_method(nstr, "members", rb_struct_s_members_m, 0);
+    rb_define_singleton_method(nstr, "inspect", rb_struct_s_inspect, 0); // FIXME: just a separate method?..
+
+    len = RARRAY_LEN(members);
+    for (i=0; i< len; i++) {
+        VALUE sym = RARRAY_AREF(members, i);
+        VALUE off = LONG2NUM(i);
+
+        define_aref_method(nstr, sym, off);
+    }
```

Comment:
As mentioned in https://bugs.ruby-lang.org/issues/16122#note-69, I think we should define `initialize` on the subclass, and not as Data#initialize.